sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -595,58 +608,65 @@ public BasicDocumentRevision call(SQLDatabase db) throws Exception {
-                options.docNumericId = potentialParent.getInternalNumericId();
-                options.parentSequence = potentialParent.getSequence();
-            } else {
-                // otherwise we are doing a 'normal' create document
-                long docNumericId = insertDocumentID(db,docId);
-                options.revId = CouchUtils.getFirstRevisionId();
-                options.docNumericId = docNumericId;
-                options.parentSequence = -1l;
-            }
+        InsertRevisionOptions options = new InsertRevisionOptions();
+        BasicDocumentRevision potentialParent = null;
 
-            options.deleted = false;
-            options.current = true;
-            options.data = body.asBytes();
-            options.available = true;
-            options.copyAttachments = false;
-            long newSequence = insertRevision(db,options);
-            if (newSequence < 0) {
-                throw new IllegalStateException("Error inserting new revision: "+options);
-            }
+        try {
+            potentialParent = this.getDocumentInQueue(db, docId, null);
+        } catch (DocumentNotFoundException e) {
+            //this is an expected exception, it just means we are
+            // resurrecting the document
+        }
 
-            BasicDocumentRevision doc = getDocumentInQueue(db, docId, options.revId);
-            documentCreated = new DocumentCreated(doc);
+        if (potentialParent != null) {
+            if (!potentialParent.isDeleted()) {
+                // current winner not deleted, can't insert
+                throw new ConflictException(String.format("Can not create new doc, document with id %s already exists "
+                        , docId));
+            }
+            // if we got here, parent rev was deleted
+            this.setCurrent(db,potentialParent, false);
+            options.revId = CouchUtils.generateNextRevisionId(potentialParent.getRevision());
+            options.docNumericId = potentialParent.getInternalNumericId();
+            options.parentSequence = potentialParent.getSequence();
+        } else {
+            // otherwise we are doing a 'normal' create document
+            long docNumericId = insertDocumentID(db,docId);
+            options.revId = CouchUtils.getFirstRevisionId();
+            options.docNumericId = docNumericId;
+            options.parentSequence = -1l;
+        }
+        options.deleted = false;
+        options.current = true;
+        options.data = body.asBytes();
+        options.available = true;
+        options.copyAttachments = false;
+        insertRevision(db, options);
 
+        try {
+            BasicDocumentRevision doc =  getDocumentInQueue(db, docId, options.revId);
             logger.finer("New document created: " + doc.toString());
-
             return doc;
+        } catch (DocumentNotFoundException e){
+            throw new RuntimeException("Programming error we just inserted it so we should have it",e);
```

Comment:
Ideally, a user should never see this exception thrown, since the insert passed we fully expect the document to be found, so the exception shouldn't be throw. However we can't swallow the exception, an error will appear somewhere else. Perhaps the message should read `Could not get newly created document with id: %s ` ?